src/core/magnetostatics/dds.cpp

```diff
@@ -30,110 +30,376 @@
+
+  /* Output iterator for the force */
+  auto p = local_particles.begin();
+
+  /* IA with local particles */
+  for (auto pi = local_posmom_begin; pi != local_posmom_end; ++pi, ++p) {
+    /* IA with own images */
+    auto fi = image_sum(
+        pi, std::next(pi), pi, with_replicas, ncut, box_l, ParticleForce{},
+        [pi](Utils::Vector3d const &rn, Utils::Vector3d const &mj) {
+          return pair_force(rn, pi->m, mj);
+        });
+
+    /* IA with other local particles */
+    auto q = std::next(p);
+    for (auto pj = std::next(pi); pj != local_posmom_end; ++pj, ++q) {
+      auto const d = (with_replicas)
+                         ? (pi->pos - pj->pos)
+                         : ::box_geo.get_mi_vector(pi->pos, pj->pos);
+
+      ParticleForce fij{};
+      ParticleForce fji{};
+      for_each_image(ncut, [&](int nx, int ny, int nz) {
+        auto const rn =
+            d + Utils::Vector3d{nx * box_l[0], ny * box_l[1], nz * box_l[2]};
+        auto const pf = pair_force(rn, pi->m, pj->m);
+        fij += pf;
+        fji.f -= pf.f;
+        /* Conservation of angular momentum mandates that
+         * 0 = t_i + r_ij x F_ij + t_j */
+        fji.torque += vector_product(pf.f, rn) - pf.torque;
+      });
+
+      fi += fij;
+      (*q)->f.f += prefactor * fji.f;
+      (*q)->f.torque += prefactor * fji.torque;
     }
+
+    (*p)->force() += prefactor * fi.f;
+    (*p)->torque() += prefactor * fi.torque;
   }
 
-  return energy;
+  /* Wait for the rest of the data to arrive */
+  boost::mpi::wait_all(reqs.begin(), reqs.end());
+
+  /* Output iterator for the force */
+  p = local_particles.begin();
+
+  /* Interaction with all the other particles */
+  for (auto pi = local_posmom_begin; pi != local_posmom_end; ++pi, ++p) {
+    // red particles
+    auto fi =
+        image_sum(all_posmom.begin(), local_posmom_begin, pi, with_replicas,
+                  ncut, box_l, ParticleForce{},
+                  [pi](Utils::Vector3d const &rn, Utils::Vector3d const &mj) {
+                    return pair_force(rn, pi->m, mj);
+                  });
+
+    // black particles
+    fi += image_sum(local_posmom_end, all_posmom.end(), pi, with_replicas, ncut,
+                    box_l, ParticleForce{},
+                    [pi](Utils::Vector3d const &rn, Utils::Vector3d const &mj) {
+                      return pair_force(rn, pi->m, mj);
+                    });
+
+    (*p)->force() += prefactor * fi.f;
+    (*p)->torque() += prefactor * fi.torque;
+  }
 }
 
-DipolarDirectSum::DipolarDirectSum(double prefactor) : prefactor{prefactor} {
-  if (n_nodes > 1) {
-    throw std::runtime_error(
-        "MPI parallelization not supported by DipolarDirectSumCpu.");
+/**
+ * @brief Calculate the interaction potential.
+ *
+ * This employs a parallel N-square loop over all particle pairs.
+ */
+double
+DipolarDirectSum::long_range_energy(ParticleRange const &particles) const {
+  auto const &box_l = ::box_geo.length();
```

Comment:
This kernel looks structurally different from the force kernel. Is there an mpi reduce somewhere else in Es whic accumulates the energy contributions from all the nodes?

Reply:
Yes, this is managed by the `Observable_stat` framework, which is always called in parallel from the corresponding observable and analysis function.